File: VB/Q139895_2/module.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraEditors
Imports DevExpress.XtraBars.Ribbon
Imports System.Windows.Forms
Imports DevExpress.XtraBars

Namespace DXSample
    Public Class Editor
        Inherits XtraForm
        Implements IItemRefreshSupport
        Public Sub New()
            MyBase.New()
            InitializeComponent()
        End Sub
        Private Sub InitializeComponent()
            Me.ribbon = New DevExpress.XtraBars.Ribbon.RibbonControl()
            Me.bbiSave = New DevExpress.XtraBars.BarButtonItem()
            Me.bbiClose = New DevExpress.XtraBars.BarButtonItem()
            Me.rpEdit = New DevExpress.XtraBars.Ribbon.RibbonPage()
            Me.rpgEdit = New DevExpress.XtraBars.Ribbon.RibbonPageGroup()
            Me.memo = New DevExpress.XtraEditors.MemoEdit()
            CType(Me.memo.Properties, System.ComponentModel.ISupportInitialize).BeginInit()
            Me.SuspendLayout()
            ' 
            ' ribbon
            ' 
            Me.ribbon.ApplicationButtonKeyTip = ""
            Me.ribbon.ApplicationIcon = Nothing
            Me.ribbon.Items.AddRange(New DevExpress.XtraBars.BarItem() {Me.bbiSave, Me.bbiClose})
            Me.ribbon.Location = New System.Drawing.Point(0, 0)
            Me.ribbon.MaxItemId = 2
            Me.ribbon.Name = "ribbon"
            Me.ribbon.Pages.AddRange(New DevExpress.XtraBars.Ribbon.RibbonPage() {Me.rpEdit})
            Me.ribbon.SelectedPage = Me.rpEdit
            Me.ribbon.Size = New System.Drawing.Size(284, 141)
            ' 
            ' bbiSave
            ' 
            Me.bbiSave.Caption = "Save"
            Me.bbiSave.Id = 0
            Me.bbiSave.Name = "bbiSave"
            Me.bbiSave.Visibility = DevExpress.XtraBars.BarItemVisibility.Never
            '			Me.bbiSave.ItemClick += New DevExpress.XtraBars.ItemClickEventHandler(Me.bbiSave_ItemClick);
            ' 
            ' bbiClose
            ' 
            Me.bbiClose.Caption = "Close"
            Me.bbiClose.Id = 1
            Me.bbiClose.Name = "bbiClose"
            '			Me.bbiClose.ItemClick += New DevExpress.XtraBars.ItemClickEventHandler(Me.bbiClose_ItemClick);
            ' 
            ' rpEdit
            ' 
            Me.rpEdit.Groups.AddRange(New DevExpress.XtraBars.Ribbon.RibbonPageGroup() {Me.rpgEdit})
            Me.rpEdit.KeyTip = ""
            Me.rpEdit.Name = "rpEdit"
            Me.rpEdit.Text = "Edit"
            ' 
            ' rpgEdit
            ' 
            Me.rpgEdit.AllowTextClipping = False
            Me.rpgEdit.ItemLinks.Add(Me.bbiSave)
            Me.rpgEdit.ItemLinks.Add(Me.bbiClose)
            Me.rpgEdit.KeyTip = ""
            Me.rpgEdit.Name = "rpgEdit"
            Me.rpgEdit.Text = "Opened"
            ' 
            ' memo
            ' 
            Me.memo.Dock = System.Windows.Forms.DockStyle.Fill
            Me.memo.Location = New System.Drawing.Point(0, 141)
            Me.memo.Name = "memo"
            Me.memo.Size = New System.Drawing.Size(284, 123)
            Me.memo.TabIndex = 1
            '			Me.memo.TextChanged += New System.EventHandler(Me.memo_TextChanged);
            ' 
            ' Module
            ' 
            Me.ClientSize = New System.Drawing.Size(284, 264)
            Me.Controls.Add(Me.memo)
            Me.Controls.Add(Me.ribbon)
            Me.Name = "Module"
            Me.Text = "untitled"
            '			Me.Load += New System.EventHandler(Me.Module_Load);
            CType(Me.memo.Properties, System.ComponentModel.ISupportInitialize).EndInit()
            Me.ResumeLayout(False)

        End Sub
        Private ribbon As RibbonControl
        Private rpEdit As RibbonPage
        Private rpgEdit As RibbonPageGroup
        Private WithEvents bbiSave As DevExpress.XtraBars.BarButtonItem
        Private WithEvents bbiClose As DevExpress.XtraBars.BarButtonItem
        Private WithEvents memo As MemoEdit

#Region "IItemRefreshSupport Members"

        Public Function GetCategory(ByVal text As String) As RibbonPageCategory Implements IItemRefreshSupport.GetCategory
            For Each category As RibbonPageCategory In ribbon.PageCategories
                If category.Name = text Then
                    Return category
                End If
            Next category
            Return Nothing
        End Function

        Public Function GetPage(ByVal text As String) As RibbonPage Implements IItemRefreshSupport.GetPage
            For Each page As RibbonPage In ribbon.Pages
                If page.Name = text Then
                    Return page
                End If
            Next page
            Return Nothing
        End Function

        Public Function GetGroup(ByVal text As String) As RibbonPageGroup Implements IItemRefreshSupport.GetGroup
            For Each page As RibbonPage In ribbon.Pages
                For Each group As RibbonPageGroup In page.Groups
                    If group.Name = text Then
                        Return group
                    End If
                Next group
            Next page
            Return Nothing
        End Function

        Public ReadOnly Property IsMerged() As Boolean Implements IItemRefreshSupport.IsMerged
            Get
                Select Case ribbon.MdiMergeStyle
                    Case RibbonMdiMergeStyle.Always
                        Return True
                    Case RibbonMdiMergeStyle.Default, RibbonMdiMergeStyle.OnlyWhenMaximized
                        If WindowState = FormWindowState.Maximized Then
                            Return True
                        Else
                            Return False
                        End If
                    Case Else
                        Return False
                End Select
            End Get
        End Property

#End Region

        Private helper As ItemRefreshHelper

        Private Sub bbiSave_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles bbiSave.ItemClick
            bbiSave.Visibility = BarItemVisibility.Never
            helper.SetItemText(rpEdit.Name, "Edit", ItemType.Page)
            helper.SetItemText(rpgEdit.Name, "Saved", ItemType.Group)
        End Sub

        Private Sub Module_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
            helper = New ItemRefreshHelper(CType(MdiParent, IItemRefreshSupport), Me)
            helper.AddItem(rpEdit.Name, rpEdit.Text)
            helper.AddItem(rpgEdit.Name, rpgEdit.Text)
        End Sub

        Private Sub bbiClose_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs) Handles bbiClose.ItemClick
            Close()
        End Sub

        Private Sub memo_TextChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles memo.TextChanged
            bbiSave.Visibility = BarItemVisibility.Always
            helper.SetItemText(rpEdit.Name, "Edit *", ItemType.Page)
            helper.SetItemText(rpgEdit.Name, String.Format("Lines: {0}", memo.Lines.Length), ItemType.Group)
        End Sub
    End Class
End Namespace

File: CS/Q139895_2/Q139895_2/module.cs
using DevExpress.XtraEditors;
using DevExpress.XtraBars.Ribbon;
using System.Windows.Forms;
using DevExpress.XtraBars;

namespace DXSample {
    public class Module : XtraForm, IItemRefreshSupport {
        public Module() : base() { InitializeComponent(); }
        private void InitializeComponent() {
            this.ribbon = new DevExpress.XtraBars.Ribbon.RibbonControl();
            this.bbiSave = new DevExpress.XtraBars.BarButtonItem();
            this.bbiClose = new DevExpress.XtraBars.BarButtonItem();
            this.rpEdit = new DevExpress.XtraBars.Ribbon.RibbonPage();
            this.rpgEdit = new DevExpress.XtraBars.Ribbon.RibbonPageGroup();
            this.memo = new DevExpress.XtraEditors.MemoEdit();
            ((System.ComponentModel.ISupportInitialize)(this.memo.Properties)).BeginInit();
            this.SuspendLayout();
            // 
            // ribbon
            // 
            this.ribbon.ApplicationButtonKeyTip = "";
            this.ribbon.ApplicationIcon = null;
            this.ribbon.Items.AddRange(new DevExpress.XtraBars.BarItem[] {
            this.bbiSave,
            this.bbiClose});
            this.ribbon.Location = new System.Drawing.Point(0, 0);
            this.ribbon.MaxItemId = 2;
            this.ribbon.Name = "ribbon";
            this.ribbon.Pages.AddRange(new DevExpress.XtraBars.Ribbon.RibbonPage[] {
            this.rpEdit});
            this.ribbon.SelectedPage = this.rpEdit;
            this.ribbon.Size = new System.Drawing.Size(284, 141);
            // 
            // bbiSave
            // 
            this.bbiSave.Caption = "Save";
            this.bbiSave.Id = 0;
            this.bbiSave.Name = "bbiSave";
            this.bbiSave.Visibility = DevExpress.XtraBars.BarItemVisibility.Never;
            this.bbiSave.ItemClick += new DevExpress.XtraBars.ItemClickEventHandler(this.bbiSave_ItemClick);
            // 
            // bbiClose
            // 
            this.bbiClose.Caption = "Close";
            this.bbiClose.Id = 1;
            this.bbiClose.Name = "bbiClose";
            this.bbiClose.ItemClick += new DevExpress.XtraBars.ItemClickEventHandler(this.bbiClose_ItemClick);
            // 
            // rpEdit
            // 
            this.rpEdit.Groups.AddRange(new DevExpress.XtraBars.Ribbon.RibbonPageGroup[] {
            this.rpgEdit});
            this.rpEdit.KeyTip = "";
            this.rpEdit.Name = "rpEdit";
            this.rpEdit.Text = "Edit";
            // 
            // rpgEdit
            // 
            this.rpgEdit.AllowTextClipping = false;
            this.rpgEdit.ItemLinks.Add(this.bbiSave);
            this.rpgEdit.ItemLinks.Add(this.bbiClose);
            this.rpgEdit.KeyTip = "";
            this.rpgEdit.Name = "rpgEdit";
            this.rpgEdit.Text = "Opened";
            // 
            // memo
            // 
            this.memo.Dock = System.Windows.Forms.DockStyle.Fill;
            this.memo.Location = new System.Drawing.Point(0, 141);
            this.memo.Name = "memo";
            this.memo.Size = new System.Drawing.Size(284, 123);
            this.memo.TabIndex = 1;
            this.memo.TextChanged += new System.EventHandler(this.memo_TextChanged);
            // 
            // Module
            // 
            this.ClientSize = new System.Drawing.Size(284, 264);
            this.Controls.Add(this.memo);
            this.Controls.Add(this.ribbon);
            this.Name = "Module";
            this.Text = "untitled";
            this.Load += new System.EventHandler(this.Module_Load);
            ((System.ComponentModel.ISupportInitialize)(this.memo.Properties)).EndInit();
            this.ResumeLayout(false);

        }
        private RibbonControl ribbon;
        private RibbonPage rpEdit;
        private RibbonPageGroup rpgEdit;
        private DevExpress.XtraBars.BarButtonItem bbiSave;
        private DevExpress.XtraBars.BarButtonItem bbiClose;
        private MemoEdit memo;

        #region IItemRefreshSupport Members

        public RibbonPageCategory GetCategory(string text) { 
            foreach (RibbonPageCategory category in ribbon.PageCategories) if (category.Name == text) return category;
            return null;
        }

        public RibbonPage GetPage(string text) { 
            foreach (RibbonPage page in ribbon.Pages) if (page.Name == text) return page;
            return null;
        }

        public RibbonPageGroup GetGroup(string text) {
            foreach (RibbonPage page in ribbon.Pages) foreach (RibbonPageGroup group in page.Groups) if (group.Name == text) return group;
            return null;
        }

        public bool IsMerged {
            get {
                switch (ribbon.MdiMergeStyle) {
                    case RibbonMdiMergeStyle.Always: return true;
                    case RibbonMdiMergeStyle.Default:
                    case RibbonMdiMergeStyle.OnlyWhenMaximized: return WindowState == FormWindowState.Maximized ? true : false;
                    default: return false;
                }
            }
        }

        #endregion

        private ItemRefreshHelper helper;

        private void bbiSave_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            bbiSave.Visibility = BarItemVisibility.Never;
            helper.SetItemText(rpEdit.Name, "Edit", ItemType.Page);
            helper.SetItemText(rpgEdit.Name, "Saved", ItemType.Group);
        }

        private void Module_Load(object sender, System.EventArgs e) {
            helper = new ItemRefreshHelper((IItemRefreshSupport)MdiParent, this);
            helper.AddItem(rpEdit.Name, rpEdit.Text);
            helper.AddItem(rpgEdit.Name, rpgEdit.Text);
        }

        private void bbiClose_ItemClick(object sender, ItemClickEventArgs e) { Close(); }

        private void memo_TextChanged(object sender, System.EventArgs e) {
            bbiSave.Visibility = BarItemVisibility.Always;
            helper.SetItemText(rpEdit.Name, "Edit *", ItemType.Page);
            helper.SetItemText(rpgEdit.Name, string.Format("Lines: {0}", memo.Lines.Length), ItemType.Group);
        }
    }
}

